### ISSUE TITLE
Enemy hitchances are not working as intended.

### ISSUE THREAD

**Issue by fuglore** [segabl/pd2-streamlined-heisting#3]
Due to a mistake OVERKILL made in CopActionShoot _get_unit_shoot_pos, the calculation that creates a new position to be used for missed shots will add the weapon's usage spread value multiplied by math.random() (so anything between 0 and 1), while code that is executed after this function applies spread regardless of the result using the full amount of the weapon's usage spread.

In order for shots to hit a player, the direction of the bullet vector has to be within 30 centimeters of the local player's camera. Because of the issue above, that second application of spread can potentially redirect the bullet back into the 30 centimeter radius.

This, obviously, shouldn't be a thing, as it invalidates the acc stat on charactertweakdata's usage presets.

There's also another separate issue with the way the miss vector is created in the first place, and it's that the positions used to create a perpendicular direction are based on the shooter's m_pos (where they're standing on, it's located at feet height), instead of using the position that the unit shoots from, which is m_head_pos (literally the center of the head).

Lastly, there's an issue when it comes to shooting other human/person targets that aren't the local player, and it's the minimum distance used for the miss + the detection/attention pos is used instead of center of mass when creating the miss vector, due to how that whole system was changed here in PD2.

This is how the code normally looks like:

```lua

local enemy_vec = temp_vec2

mvec3_set(enemy_vec, pos)
mvec3_sub(enemy_vec, self._common_data.pos)

local error_vec = Vector3()

mvec3_cross(error_vec, enemy_vec, math.UP)
mrot_axis_angle(temp_rot1, enemy_vec, math.random(360))
mvec3_rot(error_vec, temp_rot1)

local miss_min_dis = shooting_local_player and 31 or 150
local error_vec_len = miss_min_dis + w_tweak.spread * math.random() + w_tweak.miss_dis * math.random() * (1 - focus_prog)

mvec3_set_l(error_vec, error_vec_len)
mvec3_add(error_vec, pos)
mvec3_set(shoot_hist.m_last_pos, error_vec)

return error_vec
```

This is how it looks like when fixed:

```lua

local enemy_vec = temp_vec2

if shooting_local_player then
    mvec3_set(enemy_vec, pos)
elseif self._attention and self._attention.unit then
    local att_mov_ext = self._attention.unit:movement()

    if not att_mov_ext or not att_mov_ext.m_com then
        mvec3_set(enemy_vec, pos)
    else
        mvec3_set(enemy_vec, att_mov_ext:m_com())
    end
end

mvec3_sub(enemy_vec, self._shoot_from_pos) --proper shooting pos, instead of the unit's m_pos (which is where they're standing on)

local error_vec = Vector3()

mvec3_cross(error_vec, enemy_vec, math.UP)
mrot_axis_angle(temp_rot1, enemy_vec, math.random(360))
mvec3_rot(error_vec, temp_rot1)

local miss_min_dis = shooting_local_player and 31 or 120
local error_vec_len = miss_min_dis + w_tweak.spread + w_tweak.miss_dis * math.random() * (1 - focus_prog) --apply the full amount of w_tweak.spread

mvec3_set_l(error_vec, error_vec_len)
mvec3_add(error_vec, pos)

return error_vec
```

Do note that the file has received other optimizations and changes, including the fixed function above (and not doing a second application of spread if not needed in the first place), but this should be vanilla-friendly without having to go through the main update function.

EDIT: It seems one of the issues mentioned in here are actually fixed, but the common_data.m_pos thing isn't, outside of that, the rest of the information is null.

**Comment by segabl** [comment]
Thanks for the info. The spread issue is non existent since I completely removed spread from the calculation but I'll have a look at the shoot pos.

**Comment by segabl** [comment]
Addressed in 7bf7b137e653a95cd335ba9590407cac52987ae8
